Code/GraphMol/Fingerprints/MHFP.cpp

```diff
@@ -0,0 +1,308 @@
+               const unsigned char& min_radius) {
+  std::vector<std::string> shingling;
+
+  if (rings) {
+    VECT_INT_VECT rings;
+    RDKit::MolOps::symmetrizeSSSR(mol, rings);
+
+    for (size_t i = 0; i < rings.size(); i++) {
+      INT_VECT ring = rings[i];
+      std::set<unsigned int> bonds;
+
+      for (size_t j = 0; j < ring.size(); j++) {
+        int atom_idx_a = ring[j];
+
+        for (size_t k = 0; k < ring.size(); k++) {
+          int atom_idx_b = ring[k];
+
+          if (atom_idx_a != atom_idx_b) {
+            Bond* bond = mol.getBondBetweenAtoms(atom_idx_a, atom_idx_b);
+
+            if (bond != nullptr) {
+              bonds.insert(bond->getIdx());
+            }
+          }
+        }
+      }
+
+      PATH_TYPE bonds_vect(bonds.size());
+      std::copy(bonds.begin(), bonds.end(), bonds_vect.begin());
+      shingling.emplace_back(MolToSmiles(*Subgraphs::pathToSubmol(mol, bonds_vect)));
+    }
+  }
+
+  unsigned char min_radius_internal = min_radius;
+
+  if (min_radius == 0) {
+    for (auto atom : mol.atoms()) {
+      shingling.emplace_back(SmilesWrite::GetAtomSmiles(atom, false, nullptr, false, true));
+    }
+
+    min_radius_internal++;
+  }
+
+  uint32_t index = 0;
+  for (auto atom : mol.atoms()) {
+    for (unsigned char r = min_radius_internal; r < radius + 1; r++) {
+      PATH_TYPE path = findAtomEnvironmentOfRadiusN(mol, r, index);
+      INT_MAP_INT amap;
+      ROMol submol = *Subgraphs::pathToSubmol(mol, path, false, amap);
+
+      if (amap.find(index) == amap.end())
+        continue;
+
+      std::string smiles = MolToSmiles(submol, isomeric, kekulize, amap[index]);
+      
+      if (smiles != "")
+        shingling.emplace_back(smiles);
+    }
+
+    index++;
+  }
+
+  return shingling;
+}
+
+std::vector<std::string>
+MHFPEncoder::CreateShingling(std::string& smiles, 
+               const unsigned char& radius,
+               const bool& rings,
+               const bool& isomeric,
+               const bool& kekulize,
+               const unsigned char& min_radius) {
+  ROMol* mol = SmilesToMol(smiles);
+  std::vector<std::string> shingling = CreateShingling(*mol, radius, rings, isomeric, kekulize, min_radius);
+  delete mol;
+  return shingling;
+}
+
+std::vector<uint32_t>
+MHFPEncoder::Encode(ROMol& mol, 
+          const unsigned char& radius,
+          const bool& rings,
+          const bool& isomeric,
+          const bool& kekulize,
+          const unsigned char& min_radius) {
+  return FromStringArray(CreateShingling(mol, radius, rings, isomeric, kekulize, min_radius));                        
+}
+
+std::vector<std::vector<uint32_t>>
+MHFPEncoder::Encode(std::vector<ROMol>& mols, 
+          const unsigned char& radius,
+          const bool& rings,
+          const bool& isomeric,
+          const bool& kekulize,
+          const unsigned char& min_radius) {
+  size_t n = mols.size();
+  std::vector<std::vector<uint32_t>> results(n);
+  
+  #pragma omp parallel for
```

Comment:
Some of our target compilers don't support openmp, we tend to use explicit threading approaches.

Reply:
Is there a preferred way to implement a parallel_for algorithm (I couldn't find an example in the code), or should I just use `std::thread`?

Reply:
We have been using std::thread, but I would love if we could optionally use openmp.  Apple  and windows now support it so it might be viable.

@greglandrum any suggestions?

Reply:
It wouldn't be a big issue to change it (I resolved the other requests already). Doesn't msvc implement openmp since 2008? On Apple the problem is (IIRC) the bundled xcode clang?

Reply:
Greg reminded me that the RDKit code is also used in controlled environments where the need to specify the number of threads being used is quite important.  Think cluster machines and running in KNIME nodes.

Using a std::future might be relatively easy:

```
https://github.com/rdkit/rdkit/blob/d8c49e6dabfa4c48fcc58b0e8023f8cc00c00981/Code/GraphMol/ForceFieldHelpers/MMFF/testMultiThread.cpp#L75
```